File: packages/gavio-js/src/interceptors/audit/record.ts
/** AuditRecord — the immutable, per-request audit entry. */

import { createHash } from 'node:crypto'
import { PromptLineage, TokenUsage } from '../../types.js'

export const SCHEMA_VERSION = '1.0'

function sha256(text: string): string {
  return createHash('sha256').update(text, 'utf-8').digest('hex')
}

/** Deterministic JSON with keys sorted at every nesting level. */
function stableStringify(value: unknown): string {
  if (value === null || typeof value !== 'object') return JSON.stringify(value) ?? 'null'
  if (Array.isArray(value)) return `[${value.map(stableStringify).join(',')}]`
  const obj = value as Record<string, unknown>
  const parts = Object.keys(obj)
    .sort()
    .map((k) => `${JSON.stringify(k)}:${stableStringify(obj[k])}`)
  return `{${parts.join(',')}}`
}

export interface AuditRecordInit {
  traceId: string
  provider: string
  model: string
  timestampUtc: string
  parentTraceId?: string | null
  agentId?: string | null
  sessionId?: string | null
  subjectId?: string | null
  modelVersion?: string
  promptHash?: string
  responseHash?: string
  tokenUsage?: TokenUsage
  costUsd?: number
  latencyMs?: number
  piiEntityTypes?: string[]
  piiEntityCounts?: Record<string, number>
  interceptorsFired?: string[]
  cacheHit?: boolean
  cacheType?: string | null
  guardrailOutcome?: string | null
  riskScore?: number | null
  lineage?: PromptLineage | null
  previousHash?: string
  schemaVersion?: string
}

/**
 * One append-only audit entry. Carries metadata only — never raw content.
 *
 * `promptHash` / `responseHash` are SHA-256 digests so the entry is verifiable
 * without storing sensitive text. `previousHash` is reserved for the v0.2.0
 * hash-chain (F-OBS-02); empty in v0.1.0.
 */
export class AuditRecord {
  traceId: string
  provider: string
  model: string
  timestampUtc: string
  parentTraceId: string | null
  agentId: string | null
  sessionId: string | null
  subjectId: string | null
  modelVersion: string
  promptHash: string
  responseHash: string
  tokenUsage: TokenUsage
  costUsd: number
  latencyMs: number
  piiEntityTypes: string[]
  piiEntityCounts: Record<string, number>
  interceptorsFired: string[]
  cacheHit: boolean
  cacheType: string | null
  guardrailOutcome: string | null
  riskScore: number | null
  lineage: PromptLineage | null
  previousHash: string
  schemaVersion: string

  constructor(init: AuditRecordInit) {
    this.traceId = init.traceId
    this.provider = init.provider
    this.model = init.model
    this.timestampUtc = init.timestampUtc
    this.parentTraceId = init.parentTraceId ?? null
    this.agentId = init.agentId ?? null
    this.sessionId = init.sessionId ?? null
    this.subjectId = init.subjectId ?? null
    this.modelVersion = init.modelVersion ?? ''
    this.promptHash = init.promptHash ?? ''
    this.responseHash = init.responseHash ?? ''
    this.tokenUsage = init.tokenUsage ?? new TokenUsage()
    this.costUsd = init.costUsd ?? 0.0
    this.latencyMs = init.latencyMs ?? 0
    this.piiEntityTypes = init.piiEntityTypes ?? []
    this.piiEntityCounts = init.piiEntityCounts ?? {}
    this.interceptorsFired = init.interceptorsFired ?? []
    this.cacheHit = init.cacheHit ?? false
    this.cacheType = init.cacheType ?? null
    this.guardrailOutcome = init.guardrailOutcome ?? null
    this.riskScore = init.riskScore ?? null
    this.lineage = init.lineage ?? null
    this.previousHash = init.previousHash ?? ''
    this.schemaVersion = init.schemaVersion ?? SCHEMA_VERSION
  }

  static nowUtc(): string {
    return new Date().toISOString()
  }

  static hashText(text: string): string {
    return sha256(text)
  }

  toJSON(): Record<string, unknown> {
    return {
      traceId: this.traceId,
      parentTraceId: this.parentTraceId,
      agentId: this.agentId,
      sessionId: this.sessionId,
      subjectId: this.subjectId,
      provider: this.provider,
      model: this.model,
      modelVersion: this.modelVersion,
      timestampUtc: this.timestampUtc,
      promptHash: this.promptHash,
      responseHash: this.responseHash,
      tokenUsage: this.tokenUsage.toJSON(),
      costUsd: this.costUsd,
      latencyMs: this.latencyMs,
      piiEntityTypes: this.piiEntityTypes,
      piiEntityCounts: this.piiEntityCounts,
      interceptorsFired: this.interceptorsFired,
      cacheHit: this.cacheHit,
      cacheType: this.cacheType,
      guardrailOutcome: this.guardrailOutcome,
      riskScore: this.riskScore,
      lineage: this.lineage ? this.lineage.toJSON() : null,
      previousHash: this.previousHash,
      schemaVersion: this.schemaVersion,
    }
  }

  /** Stable JSON with recursively sorted keys — used for the v0.2.0 hash chain. */
  toCanonicalJson(): string {
    return stableStringify(this.toJSON())
  }

  /** Hash of this record's content — used to build the v0.2.0 chain. */
  contentHash(): string {
    return sha256(this.toCanonicalJson())
  }
}

File: packages/gavio-py/gavio/interceptors/audit/record.py
"""AuditRecord — the immutable, per-request audit entry."""

from __future__ import annotations

import hashlib
import json
from dataclasses import asdict, dataclass, field
from datetime import datetime, timezone

from ...types import PromptLineage, TokenUsage

SCHEMA_VERSION = "1.0"


def _sha256(text: str) -> str:
    return hashlib.sha256(text.encode("utf-8")).hexdigest()


@dataclass
class AuditRecord:
    """One append-only audit entry. Carries metadata only — never raw content.

    ``prompt_hash`` / ``response_hash`` are SHA-256 digests so the entry is
    verifiable without storing sensitive text. ``previous_hash`` is reserved
    for the v0.2.0 hash-chain (F-OBS-02); empty in v0.1.0.
    """

    trace_id: str
    provider: str
    model: str
    timestamp_utc: str
    parent_trace_id: str | None = None
    agent_id: str | None = None
    session_id: str | None = None
    model_version: str = ""
    prompt_hash: str = ""
    response_hash: str = ""
    token_usage: TokenUsage = field(default_factory=TokenUsage)
    cost_usd: float = 0.0
    latency_ms: int = 0
    pii_entity_types: list[str] = field(default_factory=list)
    pii_entity_counts: dict[str, int] = field(default_factory=dict)
    interceptors_fired: list[str] = field(default_factory=list)
    cache_hit: bool = False
    cache_type: str | None = None
    guardrail_outcome: str | None = None
    risk_score: float | None = None
    lineage: PromptLineage | None = None
    previous_hash: str = ""
    schema_version: str = SCHEMA_VERSION

    @staticmethod
    def now_utc() -> str:
        return datetime.now(timezone.utc).isoformat()

    @staticmethod
    def hash_text(text: str) -> str:
        return _sha256(text)

    def to_dict(self) -> dict:
        data = asdict(self)
        data["token_usage"] = self.token_usage.to_dict()
        return data

    def to_json(self) -> str:
        return json.dumps(self.to_dict(), sort_keys=True)

    def content_hash(self) -> str:
        """Hash of this record's content — used to build the v0.2.0 chain."""
        return _sha256(self.to_json())
